src/test/ui/did_you_mean/recursion_limit_deref.stderr

```diff
@@ -19,5 +19,5 @@ error[E0308]: mismatched types
    = note: expected type `&Bottom`
               found type `&Top`
 
-error: aborting due to 3 previous errors
+error: aborting due to 51 previous errors
```

Comment:
Oh boy, what happened here!? Given how much of an improvement this is, I wouldn't block on this, but it is worrisome if this PR is generating this many non-emitted diagnostics.

Reply:
I would actually consider this blocking (changing behavior isn't always wrong, but changing behavior that _no one understands_ is pretty bad).

Reply:
It turns out that [we don't emit exactly-equal (by hash) diagnostics](https://github.com/rust-lang/rust/blob/16992930835ce3376a4aaed42307726e1fc78e45/src/librustc_errors/lib.rs#L587-L590), the method-probing code [uses `Autoderef`](https://github.com/rust-lang/rust/blob/16992930835ce3376a4aaed42307726e1fc78e45/src/librustc_typeck/check/method/probe.rs#L299), and `Autoderef` [doesn't know when it's already](https://github.com/rust-lang/rust/blob/16992930835ce3376a4aaed42307726e1fc78e45/src/librustc_typeck/check/autoderef.rs#L60-L72) set the recursion-limit error.

Reply:
... or rather, it's not that `Autoderef` needs to know when it's already emitted the error, but rather that we instantiate `Autoderef` many times and hit the limit many times when searching for suggestions.